internals-js/src/operations.ts

```diff
@@ -1629,13 +1641,44 @@ export class FieldSelection extends Freezable<FieldSelection> {
         //   }
         // To do that, we can change that `equals` to `contains`, but then we should also "extract" the remainder
         // of `optimizedSelection` that isn't covered by the fragment, and that is the part slighly more involved.
-        if (optimizedSelection.equals(candidate.selectionSet)) {
+        if (optimizedSelection && optimizedSelection.equals(candidate.selectionSet)) {
           const fragmentSelection = new FragmentSpreadSelection(fieldBaseType, fragments, candidate.name);
           return new FieldSelection(this.field, selectionSetOf(fieldBaseType, fragmentSelection));
+        } else if(options?.autoFragmetize && optimizedSelection && optimizedSelection.contains(candidate.selectionSet)) {
```

Comment:
Some similar `contains` check has been added on `main`, so the patch here will only need the later part where a fragment is optimistically generated if the remaining selection is not "trivial". But I'm afraid this will require a bit of rebase, sorry.